hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/FileGroupReaderBasedMergeHandle.java

```diff
@@ -289,4 +469,141 @@ private GenericRecord convertOutput(T record) {
       return convertedRecord == null ? null : readerContext.getRecordContext().convertToAvroRecord(convertedRecord, requestedSchema.get());
     }
   }
+
+  private static class RecordLevelIndexCallback<T> implements BaseFileUpdateCallback<T> {
+    private final Map<String, HoodieRecordUpdateState> recordUpdateStateMap = new HashMap<>();
+
+    @Override
+    public void onUpdate(String recordKey, T previousRecord, T mergedRecord) {
+      this.recordUpdateStateMap.put(recordKey, HoodieRecordUpdateState.UPDATE);
```

Comment:
there is no need to use a map to keep the state, the record is handled one-by-one, just one variable here should be enough.

Reply:
I spoke with Siva about this and we can get rid of the map by directly updating the write status here similar to how the secondary index updates are handled. We can also further optimize to avoid converting to HoodieRecord for the secondary index update since that wrapper is just used for field access and we can use the RecordContext for fetching the fields.